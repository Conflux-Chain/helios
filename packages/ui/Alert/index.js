import React, {useMemo} from 'react'
import PropTypes from 'prop-types'
import ErrorFilled from './assets/ErrorFilled'
import SuccessFilled from './assets/SuccessFilled'
import WarningFilled from './assets/WarningFilled'
import InfoFilled from './assets/InfoFilled'
import Close from './assets/Close'

const typeToIcon = {
  success: <SuccessFilled className="w-4 h-4 mr-2" />,
  info: <InfoFilled className="w-4 h-4 mr-2" />,
  error: <ErrorFilled className="w-4 h-4 mr-2" />,
  warning: <WarningFilled className="w-4 h-4 mr-2" />,
}
function Alert({
  type = 'success',
  icon,
  className = '',
  width = 'w-100',
  open = false,
  closable = true,
  closeIcon,
  onClose,
  content,
  bordered = false,
}) {
  const IconComponent = typeToIcon[type]

  const style = useMemo(() => {
    if (type === 'info')
      return `bg-info-10 border ${bordered ? 'border-info' : 'border-info-10'}`
    if (type === 'warning')
      return `bg-warning-10 border ${
        bordered ? 'border-warning' : 'border-warning-10'
      }`
    if (type === 'error')
      return `bg-error-10 border ${
        bordered ? 'border-error' : 'border-error-10'
      }`
    if (type === 'success')
      return `bg-success-10 border ${
        bordered ? 'border-success' : 'border-success-10'
      }`
  }, [bordered, type])

  const contentStyle = useMemo(() => {
    if (type === 'info') return 'text-info-dark'
    if (type === 'warning') return 'text-warning-dark'
    if (type === 'error') return 'text-error-dark'
    if (type === 'success') return 'text-success-dark'
  }, [type])

  const onCloseClick = e => {
    e.stopPropagation()
    if (closable) {
      onClose && onClose(e)
    }
  }

  const closeIconComp = closeIcon
    ? React.cloneElement(closeIcon, {
        className: `w-4 h-4 text-gray-40 ${closeIcon.props.className || ''}`,
      })
    : null

  if (!open) return null

  return (
    <div
      data-testid="alert-wrapper"
      className={`flex items-center justify-between px-4 py-3 ${style} ${width} ${className}`}
    >
      <div className="flex items-center">
        {icon || IconComponent || null}
        <span className={`${contentStyle}`}>{content}</span>
      </div>
      {closable && (
        <span
          aria-hidden="true"
          data-testid="alert-close"
          onClick={e => onCloseClick(e)}
        >
          {closeIcon ? (
            closeIconComp
          ) : (
            <Close
              className="w-4 h-4 text-gray-40"
              onClick={e => onCloseClick(e)}
            />
          )}
        </span>
      )}
    </div>
  )
}

export default Alert

Alert.propTypes = {
  type: PropTypes.oneOf(['info', 'success', 'error', 'warning']),
  className: PropTypes.string,
  width: PropTypes.string,
  open: PropTypes.bool,
  closable: PropTypes.bool,
  bordered: PropTypes.bool,
  onClose: PropTypes.func,
  closeIcon: PropTypes.node,
  icon: PropTypes.node,
  content: PropTypes.oneOfType([PropTypes.string, PropTypes.node]),
}
