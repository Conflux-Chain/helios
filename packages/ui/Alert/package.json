{
  "private": true,
  "version": "0.0.0",
  "name": "@cfxjs/component-alert",
  "type": "module",
  "main": "index.js"
}
