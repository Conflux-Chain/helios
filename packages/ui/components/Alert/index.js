import React, {useMemo} from 'react'
import PropTypes from 'prop-types'
import {
  CloseOutlined,
  CloseCircleFilled,
  InfoCircleFilled,
  CheckCircleFilled,
  ExclamationCircleFilled,
} from '@fluent-wallet/component-icons'

const typeToIcon = {
  success: <CheckCircleFilled className="w-4 h-4 mr-2 text-success" />,
  info: <InfoCircleFilled className="w-4 h-4 mr-2 text-info" />,
  error: <CloseCircleFilled className="w-4 h-4 mr-2 text-error" />,
  warning: <ExclamationCircleFilled className="w-4 h-4 mr-2 text-warning" />,
}

const contentStyleObj = {
  info: 'text-info-dark',
  warning: 'text-warning-dark',
  error: 'text-error-dark',
  success: 'text-success-dark',
}
function Alert({
  type = 'success',
  icon,
  className = '',
  width = 'w-100',
  open = false,
  closable = true,
  closeIcon,
  onClose,
  content,
  bordered = false,
}) {
  const IconComponent = typeToIcon[type]

  const style = useMemo(() => {
    if (type === 'info')
      return `bg-info-10 ${bordered ? 'border-info' : 'border-info-10'}`
    if (type === 'warning')
      return `bg-warning-10 ${
        bordered ? 'border-warning' : 'border-warning-10'
      }`
    if (type === 'error')
      return `bg-error-10 ${bordered ? 'border-error' : 'border-error-10'}`
    if (type === 'success')
      return `bg-success-10 ${
        bordered ? 'border-success' : 'border-success-10'
      }`
  }, [bordered, type])

  const contentStyle = contentStyleObj[type] || ''

  const onCloseClick = e => {
    e.stopPropagation()
    if (closable) {
      onClose && onClose(e)
    }
  }

  const closeIconComp = closeIcon
    ? React.cloneElement(closeIcon, {
        className: `w-4 h-4 text-gray-40 ${closeIcon.props.className || ''}`,
      })
    : null

  if (!open) return null

  return (
    <div
      data-testid="alert-wrapper"
      className={`flex items-center justify-between px-4 py-3 border border-solid ${style} ${width} ${className}`}
    >
      <div className="flex items-start">
        <span className="shrink-0">{icon || IconComponent || null}</span>
        <span className={`${contentStyle}`}>{content}</span>
      </div>
      {closable && (
        <span
          aria-hidden="true"
          data-testid="alert-close"
          onClick={e => onCloseClick(e)}
        >
          {closeIcon ? (
            closeIconComp
          ) : (
            <CloseOutlined
              className="w-4 h-4 text-gray-40"
              onClick={e => onCloseClick(e)}
            />
          )}
        </span>
      )}
    </div>
  )
}

export default Alert

Alert.propTypes = {
  type: PropTypes.oneOf(['info', 'success', 'error', 'warning']),
  className: PropTypes.string,
  width: PropTypes.string,
  open: PropTypes.bool,
  closable: PropTypes.bool,
  bordered: PropTypes.bool,
  onClose: PropTypes.func,
  closeIcon: PropTypes.node,
  icon: PropTypes.node,
  content: PropTypes.oneOfType([PropTypes.string, PropTypes.node]),
}
