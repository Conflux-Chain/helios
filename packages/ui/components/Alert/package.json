{
  "private": true,
  "version": "0.0.2",
  "name": "@fluent-wallet/component-alert",
  "type": "module",
  "main": "index.js",
  "peerDependencies": {
    "react": "^17",
    "react-dom": "^17"
  }
}
