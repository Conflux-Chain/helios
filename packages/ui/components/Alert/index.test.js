import {render, screen, fireEvent, cleanup} from '@testing-library/react'
import {beforeEach, describe, expect, vi, it, afterEach} from 'vitest'
import Alert from './index.js'
beforeEach(cleanup)

describe('Alert', () => {
  describe('Dom structure', () => {
    let snapshot
    beforeEach(() => {
      snapshot = render(
        <Alert
          open={true}
          type="warning"
          closable={false}
          content="testValue"
          width="w-full"
        />,
      )
    })
    it('test snapshot', () => {
      expect(snapshot).toMatchSnapshot()
    })
    it('should render expected text content', () => {
      expect(screen.getByText('testValue')).toBeInTheDocument()
    })
    it('should render expected color', () => {
      expect(screen.getByTestId('alert-wrapper')).toHaveClass('bg-warning-10')
    })
    it('should render expected width', () => {
      expect(screen.getByTestId('alert-wrapper')).toHaveClass('bg-warning-10')
    })

    it('should not render close icon', () => {
      expect(screen.queryByTestId('alert-close')).not.toBeInTheDocument()
    })
  })

  describe('Dom event', () => {
    let onClose = null
    beforeEach(() => {
      onClose = vi.fn()
      render(
        <Alert
          open={true}
          type="warning"
          closable={true}
          content="testValue"
          width="w-full"
          onClose={onClose}
        />,
      )
    })
    afterEach(() => {
      onClose = null
    })
    it('should render close icon', () => {
      expect(screen.getByTestId('alert-close')).toBeInTheDocument()
    })
    it('should execute close function', () => {
      fireEvent.click(screen.getByTestId('alert-close'))
      expect(onClose).toHaveBeenCalledTimes(1)
    })
  })
})
