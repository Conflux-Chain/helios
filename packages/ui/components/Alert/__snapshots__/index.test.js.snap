// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Alert > Dom structure > test snapshot 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="flex items-center justify-between px-4 py-3 border border-solid bg-warning-10 border-warning-10 w-full "
        data-testid="alert-wrapper"
      >
        <div
          class="flex items-start"
        >
          <span
            class="shrink-0"
          >
            <svg
              class="w-4 h-4 mr-2 text-warning"
              fill="none"
              height="24"
              viewBox="0 0 24 24"
              width="24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M12 2C6.477 2 2 6.477 2 12s4.477 10 10 10 10-4.477 10-10S17.523 2 12 2z"
                fill="currentColor"
              />
              <path
                d="M12 8v4M12 16h-.01"
                stroke="#fff"
                stroke-linecap="round"
                stroke-linejoin="round"
                stroke-width="2"
              />
            </svg>
          </span>
          <span
            class="text-warning-dark"
          >
            testValue
          </span>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="flex items-center justify-between px-4 py-3 border border-solid bg-warning-10 border-warning-10 w-full "
      data-testid="alert-wrapper"
    >
      <div
        class="flex items-start"
      >
        <span
          class="shrink-0"
        >
          <svg
            class="w-4 h-4 mr-2 text-warning"
            fill="none"
            height="24"
            viewBox="0 0 24 24"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M12 2C6.477 2 2 6.477 2 12s4.477 10 10 10 10-4.477 10-10S17.523 2 12 2z"
              fill="currentColor"
            />
            <path
              d="M12 8v4M12 16h-.01"
              stroke="#fff"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
            />
          </svg>
        </span>
        <span
          class="text-warning-dark"
        >
          testValue
        </span>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
